### DISCUSSION TITLE
bug: TRPCClientError data prop not typed

### DISCUSSION THREAD

**Discussion by austins** [trpc/trpc#2464] [Q&A] [answered]
### Provide environment information

```
  System:
    OS: Windows 10 10.0.22000
    CPU: (24) x64 AMD Ryzen 9 3900X 12-Core Processor
    Memory: 3.32 GB / 15.93 GB
  Binaries:
    Node: 16.16.0 - ~\scoop\apps\nodejs-lts\current\node.EXE
    npm: 8.15.1 - ~\scoop\apps\nodejs-lts\current\bin\npm.CMD
  Browsers:
    Edge: Spartan (44.22000.120.0), Chromium (104.0.1293.47)
    Internet Explorer: 11.0.22000.120
  npmPackages:
    @trpc/client: ^10.0.0-alpha.45 => 10.0.0-alpha.45
    @trpc/server: ^10.0.0-alpha.45 => 10.0.0-alpha.45
```

### Describe the bug

This is on experimental version `10.0.0-alpha.45`

TRPCClientError's data property is of `any` type, but it should be typed to `DefaultErrorData`, which will type the values `code`,  `httpStatus`, `stack`, and `path`.

### To reproduce

In query procedure:
`throw new TRPCError({ code: "NOT_FOUND", message: "Article not found." });`

Client call:
```
        try {
            const article = await trpcClient(fetch).article.get.query({ slug: "slug-that-does-not-exist" });
        } catch (error) {
            if (error instanceof TRPCClientError) {
                return { status: error.data.httpStatus, error }; // Typescript error: "Unsafe member access .httpStatus on an `any` value."
            }
        }
```

### Additional information

Workaround: `(error.data as DefaultErrorData)`

I'm working on setting up the Fetch adapter with SvelteKit. As the response is not available in the context prior to resolving unlike Next.js, the error has to be caught in the page Load function and converted to the correct status code for SvelteKit to handle the error. As an aside, SvelteKit does plan to make substantial changes around pages and Load function in the future, so this solution might change.

### 👨‍👧‍👦 Contributing

- [ ] 🙋‍♂️ Yes, I'd be down to file a PR fixing this bug!

**Comment by KATT** [comment, marked as answer]
You need a type guard:

```ts

function isTRPCClientError(
  cause: unknown,
): cause is TRPCClientError<AppRouter> {
  if (cause instanceof TRPCClientError) {
    return true;
  }
  return false;
}
```

This should infer it properly when using it:

```ts
try {
  const helloResponse = await proxy.greeting.greeting.query({
    name: 'world',
  });
  console.log('helloResponse', helloResponse);
} catch (cause) {
  if (isTRPCClientError(cause)) {
    console.log(cause.shape);
  }
}
```


Adding tests right now to assert that this works as expected.

**Reply by KATT** [reply]
Upgrade to `10.0.0-alpha.47` and use a type guard as shown above to fix

**Reply by austins** [reply]
Thank you.

**Reply by austins** [reply]
Linking your PR for reference: #2467

**Reply by KATT** [reply]
https://alpha.trpc.io/docs/infer-types#infer-trpclienterrors-based-on-your-router

**Reply by austins** [reply]
The doc is missing the generic in `isTRPCClientError<TRouter extends AnyRouter>`.

It can also be simplified to a direct return, but that is up to preference:
```
export function isTRPCClientError<TRouter extends AnyRouter>(cause: unknown): cause is TRPCClientError<TRouter> {
    return cause instanceof TRPCClientError;
}
```

**Reply by austins** [reply]
PR #2474
